nifi-nar-bundles/nifi-iceberg-bundle/nifi-iceberg-processors/src/main/java/org/apache/nifi/processors/iceberg/converter/GenericDataConverters.java

```diff
@@ -47,45 +51,106 @@
  */
 public class GenericDataConverters {
 
-    static class SameTypeConverter extends DataConverter<Object, Object> {
+    static class PrimitiveTypeConverter extends DataConverter<Object, Object> {
+        final Type.PrimitiveType targetType;
+        final DataType sourceType;
+
+        public PrimitiveTypeConverter(final Type.PrimitiveType type, final DataType dataType) {
+            targetType = type;
+            sourceType = dataType;
+        }
 
         @Override
         public Object convert(Object data) {
-            return data;
+            switch (targetType.typeId()) {
+                case BOOLEAN:
+                    return DataTypeUtils.convertType(data, RecordFieldType.BOOLEAN.getDataType(), null);
+                case INTEGER:
+                    return DataTypeUtils.convertType(data, RecordFieldType.INT.getDataType(), null);
+                case LONG:
+                    return DataTypeUtils.convertType(data, RecordFieldType.LONG.getDataType(), null);
+                case FLOAT:
+                    return DataTypeUtils.convertType(data, RecordFieldType.FLOAT.getDataType(), null);
+                case DOUBLE:
+                    return DataTypeUtils.convertType(data, RecordFieldType.DOUBLE.getDataType(), null);
+                case DATE:
+                    return DataTypeUtils.toLocalDate(data, () -> StringUtils.isEmpty(sourceType.getFormat())
```

Comment:
The `isEmpty` seems to be unnecessary here since  the `getDateTimeFormatter` checks null on the `format` parameter.